src/raft.c

```diff
@@ -195,21 +388,39 @@ static void executeLogEntry(RedisRaftCtx *rr, raft_entry_t *entry, raft_index_t
 
     RaftReq *req = entry->user_data;
 
+    RedisModuleCtx *reply_ctx;
+    RedisModuleCtx *ctx;
+    RaftRedisCommandArray *cmds;
+    RaftRedisCommandArray tmp = {0};
+
     if (req) {
-        executeRaftRedisCommandArray(&req->r.redis.cmds, req->ctx, req->ctx);
-        entryDetachRaftReq(rr, entry);
-        RaftReqFree(req);
+        cmds = &req->r.redis.cmds;
+        ctx = req->ctx;
+        reply_ctx = req->ctx;
     } else {
-        RaftRedisCommandArray tmp = {0};
-
         if (RaftRedisCommandArrayDeserialize(&tmp,
                                              entry->data,
                                              entry->data_len) != RR_OK) {
             PANIC("Invalid Raft entry");
         }
 
-        executeRaftRedisCommandArray(&tmp, rr->ctx, NULL);
-        RaftRedisCommandArrayFree(&tmp);
+        cmds = &tmp;
+        ctx = redis_raft.ctx;
+        reply_ctx = NULL;
+    }
+
+    HandleAsking(cmds);
+
+    /* When we're in cluster mode, go through handleSharding. This will perform
+     * hash slot validation and return an error / redirection if necessary. We do this
+     * before checkLeader() to avoid multiple redirect hops.
+     */
+    if (rr->config->sharding && handleSharding(rr, req, cmds) != RR_OK) {
```

Comment:
I think we shouldn't rely on config at this point. I've forgotten completely, why do we have this config anyway? :)

Reply:
sharding enabled prevents cross slot, while sharding disabled allows cross slots in a multi/multi key command.

so if running a single redis raft cluster, might want sharding disabled for more flexibility (i.e. normal single redis server behavior).

the only way to handle this is to put the behavior I previously had back in, i.e. expanding the log format a little to calculate the slot value before append, so each node will behave the same and not be based on configuration.

Reply:
Hmm, okay, I see.

There might be a few ways to solve this. Just thinking out loud: 

1 - One way might be having that config in the state machine. As we currently have "sharding info" in the state machine, can we use that somehow? Like can we decide if sharding is enabled or not from that?. Personally, I need sometime just to go over code, can't think right now :) 

2- Your suggestion is another option. Just writing down:
- Leader computes the slot, if sharding is enabled rejects cross slot violation before appending
- If sharding is disabled, it sets `slot` as `-1` to tell state machine skip validating `slot` against slot mapping.
- State machine will get "slot" of command array as `-1` or `some real value`. If slot is `-1`, we skip validation slot mapping as these are cross slot commands and sharding is disabled. If it's a "real value", we'll check sharding mapping.

According to this, leader computes `slot`, and then sets `slot` in RaftReq and writes it to the log. On state machine, if RaftReq exists, we use slot from it, otherwise deserialize entry. Just a minor addition to this, what if we skip writing `slot` to the log but still set `slot` in RaftReq. If RaftReq exists, state machine will use from it. If it doesn't exist, `slot` value is computed after deserializing. We may detect cross slot violation at this point but this is only because sharding was disabled and this log entry was allowed. In this case, same as before, we can skip validating slot. 

Just thinking if there is a simple way we can avoid writing `slot` to the log. (For performance reasons).

Reply:
so this is my conception.  

if we have a single shard and that shard is the full slot range, then we don't have to test sharding enabled config.

yes, this could conceptually be reconfigured later, but at this point in time its the equivalent of just a single shard cluster so might as well treat it as such.

thoughts?  @yossi @ozan?

Reply:
basically something alone the lines of

```
bool isSharding(RedisRaftCtx *rr) {
    ShardingInfo *si = rr->sharding_info;
    if (si->shard_groups_num != 1) {
        return false;
    }
    
    RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(si->shard_group_map, "^", NULL, 0);

    char *key;
    size_t key_len;
    ShardGroup *sg;

    key = RedisModule_DictNextC(iter, &key_len, (void **) &sg)) != NULL);
    RedisModule_DictIteratorStop(iter);
    
    if (!sg->local || sg->slot_ranges_num != 1 || 
        sg->slot_ranges[0].start_slot != 0 || sg->slot_ranges[0].end_slot != 16383 ||
        sg->slot_ranges[0].type != SLOTRANGE_TYPE_STABLE) {
        return false
    }
    
    return true;
}
```